### PR TITLE
Made several changes to the pd_workshop exit survey emails:

### PR DESCRIPTION
**Pull request by aoby**
1. Added CSF certificate
2. Send exit survey to all who joined section, rather than just those
   marked attended.
3. Added survey_sent_at and survey_id (for completed survey) fields to
   pd_enrollment
4. Added oneoff script to back populate survey fields and send missing
   emails from before the queueing was set up.

Note - the annotations in pd/enrollment.rb for `code` and `school` reordered. This is the correct order (it's unchanged in schema.rb). The old ordering was leftover from an earlier schema mixup that's been fixed.

### CODE REVIEW

File: dashboard/app/models/pd/enrollment.rb [hunk 1]
```diff
@@ -8,16 +8,19 @@
 #  email              :string(255)      not null
 #  created_at         :datetime
 #  updated_at         :datetime
-#  code               :string(255)
 #  school             :string(255)
+#  code               :string(255)
```
Comment (ashercodeorg) [inline dashboard/app/models/pd/enrollment.rb]: Why is the relative position of `code` changing as part of this PR?

File: dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class AddPdEnrollmentSurveyFields < ActiveRecord::Migration
+  def change
+    add_column :pd_enrollments, :survey_sent_at, :datetime
+    add_column :pd_enrollments, :survey_id, :integer
+
+    # This index should have been here all along
```
Comment (ashercodeorg) [inline dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb]: Ideally, the index would be a separate migration in a separate PR.

File: dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class AddPdEnrollmentSurveyFields < ActiveRecord::Migration
+  def change
+    add_column :pd_enrollments, :survey_sent_at, :datetime
+    add_column :pd_enrollments, :survey_id, :integer
```
Comment (ashercodeorg) [inline dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb]: Mind adding a quick comment describing the data these columns will hold?

Comment (ashercodeorg) [inline dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb]: Does it make sense to have `survey_id` before `survey_sent_at`?

File: dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+    # This index should have been here all along
+    add_index :pd_enrollments, :code, unique: true
```
Comment (ashercodeorg) [inline dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb]: Note that the uniqueness is currently valid in `production`.

```
SELECT cnt, COUNT(0)
FROM (
  SELECT code, COUNT(0) AS cnt
  FROM pd_enrollments 
  GROUP BY code
) AS t
GROUP BY cnt;
+-----+----------+
| cnt | COUNT(0) |
+-----+----------+
|   1 |     1854 |
+-----+----------+
```

File: dashboard/test/models/pd/workshop_test.rb [hunk 5]
```diff
@@ -252,4 +260,12 @@ def session_on(day_offset, start_offset, end_offset)
   def today
     Date.today.in_time_zone
   end
+
+  def create_teacher(enrolled: false, in_section: false, attended: false)
```
Comment (ashercodeorg) [inline dashboard/test/models/pd/workshop_test.rb]: Does this belong in `factories.rb` for reuse?

File: pegasus/forms/pd_workshop_survey.rb [hunk 6]
```diff
@@ -76,6 +76,14 @@ def self.normalize(data)
     result
   end
 
+  def self.process_(form)
+    # Save this form id in the relevant dashboard pd_enrollment row
+    id = form[:id]
+    data = JSON.load(row[:data])
+    enrollment_id = data['enrollment_id_i']
+    DASHBOARD_DB[:pd_enrollments][enrollment_id].update(survey_id: id)
```
Comment (ashercodeorg) [inline pegasus/forms/pd_workshop_survey.rb]: This is the same as `DASHBOARD_DB[:pd_enrollments].find_by_id(enrollment_id).update(survey_id: id)`?

File: dashboard/test/models/pd/workshop_test.rb [hunk 7]
```diff
@@ -189,37 +189,45 @@ class Pd::WorkshopTest < ActiveSupport::TestCase
   end
 
   # Tests the following:
-  #   1. enrolled_not_attended teacher does not get an email
-  #   2. enrolled_attendee and unenrolled_attendee each get an email
+  #   1. enrolled-only teacher does not get an email
+  #   2. All teachers in the section get an email
   #   3. accidental_student_attendee is converted to a teacher account and gets an email
```
Comment (ashercodeorg) [inline dashboard/test/models/pd/workshop_test.rb]: Is there a reason this test isn't three tests, each testing one thing? Consider refactoring in a followup PR.

File: dashboard/app/mailers/pd/workshop_mailer.rb [hunk 8]
```diff
@@ -87,22 +87,34 @@ def teacher_enrollment_reminder(enrollment)
...
   end
 
-  # TODO: Make sure every teacher has an enrollment
   def exit_survey(workshop, teacher, enrollment)
     @workshop = workshop
     @teacher = teacher
     @is_first_workshop = Pd::Workshop.attended_by(teacher).in_state(Pd::Workshop::STATE_ENDED).count == 1
 
-    @survey_url = CDO.code_org_url "/pd-workshop-survey/#{enrollment.code}"
+    @survey_url = CDO.code_org_url "/pd-workshop-survey/#{enrollment.code}", 'https:'
```
Comment (ashercodeorg) [inline dashboard/app/mailers/pd/workshop_mailer.rb:95]: Is this variable used anywhere?

File: dashboard/app/mailers/pd/workshop_mailer.rb [hunk 9]
```diff
@@ -87,22 +87,34 @@ def teacher_enrollment_reminder(enrollment)
...
 
+  def generate_csf_certificate
+    image = create_certificate_image2(
+      dashboard_dir('app', 'assets', 'images', 'pd_workshop_certificate_csf.png'),
+      @teacher.name || @teacher.email,
+      y: 444,
+      height: 100,
+    )
+    image.format = 'jpg'
+    Base64.encode64(image.to_blob)
```
Comment (ashercodeorg) [inline dashboard/app/mailers/pd/workshop_mailer.rb:115]: Should this be [Base64.urlsafe_encode64](http://apidock.com/ruby/v1_9_3_392/Base64/urlsafe_encode64)?

Granted, if I'm understanding stuff correctly, this sanitizes the teacher's own input.

File: dashboard/app/mailers/pd/workshop_mailer.rb (hunk 8 above)
Comment (aoby) [inline dashboard/app/mailers/pd/workshop_mailer.rb:95]: Yes, in the [exit_survey](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/views/pd/workshop_mailer/exit_survey.html.haml#L25) mail view.

File: dashboard/app/models/pd/enrollment.rb (hunk 1 above)
Comment (aoby) [inline dashboard/app/models/pd/enrollment.rb]: It's not actually changing position in the schema. This annotation was incorrect. Since I re-annotated, it switched back to the actual ordering. It was switched accidentally in https://github.com/code-dot-org/code-dot-org/pull/8394 and fixed (in the schema, not the annotation) in https://github.com/code-dot-org/code-dot-org/commit/5f9894e1391854ce7e495f571c5ad1c2b27b61b3

File: dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb (hunk 3 above)
Comment (aoby) [inline dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb]: The survey id is for a completed survey, so it makes sense to go after. The survey itself it keyed on the enrollment code so we already have that data. I'll add comments and change `survey_id` to `completed_survey_id`.

File: dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb (hunk 2 above)
Comment (aoby) [inline dashboard/db/migrate/20160630000000_add_pd_enrollment_survey_fields.rb]: Ok, I'l split it out. Any idea why it wasn't there in the first place, after `add_column :pd_enrollments, :code, :string, index: true, unique: true` in the original [migration](https://github.com/code-dot-org/code-dot-org/blob/af9eecff871ed981944c4d7ed9dcefabb0197af2/dashboard/db/migrate/20160510000000_add_district_school_and_code_to_enrollment.rb#L5-L5)? I would have thought that would create the index along with the column.

File: dashboard/test/models/pd/workshop_test.rb (hunk 7 above)
Comment (aoby) [inline dashboard/test/models/pd/workshop_test.rb]: Good suggestion. I probably did it this way initially just to make sure it works with a set of enrollments not just individuals. Separate tests would be easier to read and to debug if they ever fail. I'll split it up.

File: pegasus/forms/pd_workshop_survey.rb (hunk 6 above)
Comment (aoby) [inline pegasus/forms/pd_workshop_survey.rb]: Yes. It's the sequel equivalent of ActiveRecord's `find_by_id`. See http://sequel.jeremyevans.net/rdoc/files/doc/querying_rdoc.html#label-Retrieving+a+Single+Object

But I realized I did it wrong. It's supposed to be `DASHBOARD_DB[:pd_enrollments][id: enrollment_id]...` or `DASHBOARD_DB[:pd_enrollments].where(id: enrollment_id).first`

File: dashboard/app/mailers/pd/workshop_mailer.rb (hunk 9 above)
Comment (aoby) [inline dashboard/app/mailers/pd/workshop_mailer.rb:115]: The teacher-supplied name is written as-is onto the certificate image in `create_certificate_image2`. Then `image.to_blob` is dumping the image binary, and `encode64` is encoding the resulting image to be used as an email attachment. Since it's sent as a MIME attachment and not used as part of a url, it doesn't need to be url-safe. Also note this was basically copied from the old [certificate email](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/emails/professional-development-workshop-section-receipt.md).

File: pegasus/forms/pd_workshop_survey.rb [hunk 10]
```diff
@@ -76,6 +76,14 @@ def self.normalize(data)
     result
   end
 
+  def self.process_(form)
+    # Save this form id in the relevant dashboard pd_enrollment row
+    id = form[:id]
+    data = JSON.load(row[:data])
+    enrollment_id = data['enrollment_id_i']
+    DASHBOARD_DB[:pd_enrollments][id: enrollment_id].update(survey_id: id)
```
Comment (ashercodeorg) [inline pegasus/forms/pd_workshop_survey.rb:84]: Should this be `update!`?

**Comment by ashercodeorg** [comment]
LGTM. In the future, I'd ask that migrations be placed in a separate PR from the PR that uses them. Otherwise, if the migration fails (an earlier migration fails), the dependent code is deployed without the migration.

File: pegasus/forms/pd_workshop_survey.rb (hunk 10 above)
Comment (aoby) [inline pegasus/forms/pd_workshop_survey.rb:84]: Unfortunately [sequel](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Dataset.html#method-i-update) doesn't have the bang methods.
